File: shared/hidrobart-auth.ts
/**
 * Hidrobart Auth SDK — TypeScript/Next.js
 * Copiar este archivo a cualquier app Next.js para integrarse con el login centralizado.
 *
 * USO:
 *   import { createHidrobartAuth } from '@/lib/hidrobart-auth'
 *   const auth = createHidrobartAuth({ authApiUrl: process.env.AUTH_API_URL })
 *   const result = await auth.validateToken(token)
 */

export interface HidrobartRoles {
  org: string[];
  functional: string[];
  process: Record<string, string[]>;
  tenant?: string;
}

export interface HidrobartUser {
  id: string;
  email: string;
  name: string;
  display_name?: string;
  job_title?: string;
  department?: string;
  tenant: string;
  domain: string;
  roles: HidrobartRoles;
}

export interface ValidateResult {
  valid: boolean;
  user?: HidrobartUser;
  roles?: HidrobartRoles;
  error?: string;
}

export interface AuthConfig {
  authApiUrl: string;
  loginUrl?: string;
  timeoutMs?: number;
}

export class HidrobartAuth {
  private config: Required<AuthConfig>;

  constructor(config: AuthConfig) {
    this.config = {
      authApiUrl: config.authApiUrl.replace(/\/$/, ""),
      loginUrl: config.loginUrl || "https://login.hidrobart.com",
      timeoutMs: config.timeoutMs || 5000,
    };
  }

  /**
   * Valida un token contra el servicio centralizado.
   * Llamar en cada request protegido.
   */
  async validateToken(token: string): Promise<ValidateResult> {
    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), this.config.timeoutMs);

      const res = await fetch(`${this.config.authApiUrl}/auth/validate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ token }),
        signal: controller.signal,
      });

      clearTimeout(timeout);

      if (!res.ok) {
        return { valid: false, error: `Auth service error: ${res.status}` };
      }

      return await res.json();
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        return { valid: false, error: "Auth service timeout" };
      }
      return { valid: false, error: "Auth service unreachable" };
    }
  }

  /**
   * Obtiene perfil del usuario usando su access token.
   */
  async getMe(token: string): Promise<HidrobartUser | null> {
    try {
      const res = await fetch(`${this.config.authApiUrl}/auth/me`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      if (!res.ok) return null;
      return await res.json();
    } catch {
      return null;
    }
  }

  /**
   * Verifica permiso específico.
   * Ejemplo: auth.hasPermission(roles, "hidroplus", "aprobar")
   */
  hasPermission(roles: HidrobartRoles, module: string, permission: string): boolean {
    const perms = roles.process?.[module] || [];
    return perms.includes(permission);
  }

  isAdmin(roles: HidrobartRoles): boolean {
    return roles.org.includes("SuperAdmin") || roles.org.includes("Admin");
  }

  isManager(roles: HidrobartRoles): boolean {
    return ["SuperAdmin", "Admin", "Manager"].some((r) => roles.org.includes(r));
  }

  /**
   * URL a la que redirigir para login (con return URL).
   */
  getLoginUrl(returnTo?: string): string {
    const base = this.config.loginUrl;
    if (returnTo) {
      return `${base}/login?callbackUrl=${encodeURIComponent(returnTo)}`;
    }
    return `${base}/login`;
  }
}

export function createHidrobartAuth(config: AuthConfig): HidrobartAuth {
  return new HidrobartAuth(config);
}

// ── Helper para Next.js middleware ───────────────────────────────────────────

/**
 * Ejemplo de uso en middleware.ts de una app protegida:
 *
 * import { validateHidrobartSession } from '@/lib/hidrobart-auth'
 *
 * export async function middleware(request: NextRequest) {
 *   const token = request.cookies.get('hidrobart-token')?.value
 *     || request.headers.get('authorization')?.replace('Bearer ', '')
 *
 *   if (!token) {
 *     return NextResponse.redirect(new URL('/login', AUTH_URL))
 *   }
 *
 *   const result = await validateHidrobartSession(token)
 *   if (!result.valid) {
 *     return NextResponse.redirect(new URL('/login', AUTH_URL))
 *   }
 *
 *   // Pasar info del usuario a los headers
 *   const response = NextResponse.next()
 *   response.headers.set('x-user-id', result.user?.id || '')
 *   response.headers.set('x-user-email', result.user?.email || '')
 *   return response
 * }
 */
export async function validateHidrobartSession(
  token: string,
  authApiUrl?: string
): Promise<ValidateResult> {
  const url = authApiUrl || process.env.AUTH_API_URL || "http://localhost:8000";
  const auth = createHidrobartAuth({ authApiUrl: url });
  return auth.validateToken(token);
}

File: shared/hidrobart_auth.py
"""
Hidrobart Auth SDK — Python/FastAPI
Copiar este archivo a cualquier app FastAPI para integrarse con el login centralizado.

USO:
    from hidrobart_auth import HidrobartAuthDep, require_permission

    @app.get("/mi-ruta")
    async def mi_ruta(user = Depends(HidrobartAuthDep)):
        return {"mensaje": f"Hola {user.email}"}

    @app.post("/aprobar")
    async def aprobar(user = Depends(require_permission("hidroplus", "aprobar"))):
        return {"ok": True}
"""
import httpx
import logging
from dataclasses import dataclass, field
from typing import Optional
from functools import lru_cache
from fastapi import Depends, HTTPException, Request
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials

logger = logging.getLogger(__name__)


@dataclass
class HidrobartRoles:
    org: list[str] = field(default_factory=list)
    functional: list[str] = field(default_factory=list)
    process: dict[str, list[str]] = field(default_factory=dict)
    tenant: str = ""


@dataclass
class HidrobartUser:
    id: str
    email: str
    name: str
    tenant: str
    domain: str
    roles: HidrobartRoles
    job_title: str = ""
    department: str = ""


class HidrobartAuthClient:
    def __init__(self, auth_api_url: str, timeout: float = 5.0):
        self.auth_api_url = auth_api_url.rstrip("/")
        self.timeout = timeout

    async def validate_token(self, token: str) -> Optional[HidrobartUser]:
        """Valida token contra el servicio central de auth."""
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                resp = await client.post(
                    f"{self.auth_api_url}/auth/validate",
                    json={"token": token},
                )
                if resp.status_code != 200:
                    return None
                data = resp.json()
                if not data.get("valid"):
                    return None

                user_data = data.get("user", {})
                roles_data = data.get("roles", {})

                return HidrobartUser(
                    id=user_data.get("id", ""),
                    email=user_data.get("email", ""),
                    name=user_data.get("name", ""),
                    tenant=user_data.get("tenant", ""),
                    domain=user_data.get("domain", ""),
                    job_title=user_data.get("job_title", ""),
                    department=user_data.get("department", ""),
                    roles=HidrobartRoles(
                        org=roles_data.get("org", []),
                        functional=roles_data.get("functional", []),
                        process=roles_data.get("process", {}),
                        tenant=roles_data.get("tenant", ""),
                    ),
                )
        except Exception as e:
            logger.warning(f"Auth validation error: {e}")
            return None

    def has_permission(self, user: HidrobartUser, module: str, permission: str) -> bool:
        return permission in user.roles.process.get(module, [])

    def is_admin(self, user: HidrobartUser) -> bool:
        return any(r in user.roles.org for r in ["SuperAdmin", "Admin"])

    def is_manager(self, user: HidrobartUser) -> bool:
        return any(r in user.roles.org for r in ["SuperAdmin", "Admin", "Manager"])


# ── FastAPI Dependencies ──────────────────────────────────────────────────────

import os

_auth_client: Optional[HidrobartAuthClient] = None

def get_auth_client() -> HidrobartAuthClient:
    global _auth_client
    if _auth_client is None:
        url = os.getenv("AUTH_API_URL", "http://localhost:8000")
        _auth_client = HidrobartAuthClient(auth_api_url=url)
    return _auth_client


bearer_scheme = HTTPBearer(auto_error=False)


async def get_current_user(
    credentials: Optional[HTTPAuthorizationCredentials] = Depends(bearer_scheme),
    auth_client: HidrobartAuthClient = Depends(get_auth_client),
) -> HidrobartUser:
    """Dependency: obtiene y valida el usuario del Bearer token."""
    if not credentials:
        raise HTTPException(status_code=401, detail="Token de autenticación requerido")

    user = await auth_client.validate_token(credentials.credentials)
    if not user:
        raise HTTPException(status_code=401, detail="Token inválido o sesión expirada")

    return user


# Alias
HidrobartAuthDep = Depends(get_current_user)


def require_permission(module: str, permission: str):
    """
    Dependency factory: requiere un permiso específico.

    Ejemplo:
        @app.get("/ruta")
        async def ruta(user = Depends(require_permission("hidroplus", "aprobar"))):
            ...
    """
    async def _dep(
        user: HidrobartUser = Depends(get_current_user),
        auth_client: HidrobartAuthClient = Depends(get_auth_client),
    ) -> HidrobartUser:
        if not auth_client.has_permission(user, module, permission):
            raise HTTPException(
                status_code=403,
                detail=f"Sin permiso '{permission}' en módulo '{module}'",
            )
        return user

    return Depends(_dep)


def require_admin():
    """Dependency: requiere rol Admin o SuperAdmin."""
    async def _dep(
        user: HidrobartUser = Depends(get_current_user),
        auth_client: HidrobartAuthClient = Depends(get_auth_client),
    ) -> HidrobartUser:
        if not auth_client.is_admin(user):
            raise HTTPException(status_code=403, detail="Acceso restringido a administradores")
        return user

    return Depends(_dep)
